lineairdb_transaction.hh

```diff
@@ -0,0 +1,52 @@
+#include <lineairdb/lineairdb.h>
+#include "sql/handler.h" /* handler */
+#include "mysql/plugin.h"
+#include "sql/sql_class.h"
+
+/**
+ * @brief 
+ * Wrapper of LineairDB::Transaction
+ * Takes care of registering a transaction to MySQL core
+ * Lifetime of this class equals the lifetime of the transaction
+ * The instance of this class must be deleted after `end_transaction`
```

Comment:
Sorry this comment was inaccurate. 
What I meant was the pointer to an instance of this class must be `delete`d after `end_transaction`.
In other words, "be careful of memory leak".
To be more exact, I want to enforce `cleanup_committed_tx` after `lineairdb_commit` or lineairdb_abort`.
Should I write:
`The pointer to an instance of this class must be `delete`d after `end_transaction`.
For example, `cleanup_committed_tx` is called after `lineairdb_commit` or lineairdb_abort` to avoid memory leak.`?

Reply:
> The pointer to an instance of this class must be `delete`d after `end_transaction`. For example, `cleanup_committed_tx` is called after `lineairdb_commit` or lineairdb_abort` to avoid memory leak.?

I think it is a good fix.
I also suggest adding an assertion. (I'm sorry if you have already done that.)

Besides, another solution... How about calling `cleanup_committed_tx ` in the first line of destructor automatically when `delete` is called?

Reply:
I wasn't aware that you can `delete` the pointer to self in the destructor.
I will check it out.

Reply:
I am not sure how to `delete` pointer to self in destructor.
>Besides, another solution... How about calling cleanup_committed_tx  in the first line of destructor automatically when delete is called?

Am I misunderstanding your suggestion?

Reply:
I understood that you want to match the lifetime of `LineairDBTransaction` and its member variable `tx`, (and thereby eliminate memory leaks).
I thought we could solve it by implementing a destructor.

My understanding of the comment :
```c++
auto* foo = new LineairDBTransaction();

// NG:
delete foo;

// OK:
foo->end_transaction();
delete foo;
```

So what I suggested:

````c++
// NG: default destructor does not delete its member pointer
~LineairDBTransaction() = default;

// OK: implement `delete` manually:
~LineairDBTransaction() {
     delete tx;
     // or invoke cleanup_committed_tx() here;
}

// Then you can delete foo directly without cleanup
auto* foo = new LineairDBTransaction();
delete foo; // `delete tx` is invoked at this time
```

Reply:
I find this implementation (callling `end_transaction()` in LineairDBTransaction's destructor) confusing because then the user will not know when the transaction has committed.

```
static int lineairdb_commit(handlerton *hton, THD *thd, bool shouldTerminate) {
  if (shouldTerminate == false) return 0;
  LineairDBTransaction *&tx = *reinterpret_cast<LineairDBTransaction**>(thd_ha_data(thd, hton));

  assert(tx != nullptr);
  
  delete tx;     // did the transaction commit here???
  tx = nullptr;
  
  return 0;
}
```

Ideally I want to `delete` LineairDBTransaction pointer (not LineairDB::Transaction) when I call `end_transaction`, but that seems impossible. Hence I created the function `cleanup_committed_tx` to make it clear that we need to `delete` LineairDBTransaction pointer after termination (I will change the name of the function to `cleanup_terminated_tx`).

Reply:
I think I finally understand. Sorry for repeating misleading comments.

However, I still do not understand this comment:
> Ideally I want to delete LineairDBTransaction pointer (not LineairDB::Transaction) when I call end_transaction, but that seems impossible.

I think we can do it and could delete the function `cleanup_*_tx()`.
We can invoke `delete(this)`. Is it not enough?
```c++
void LineairDBTransaction::end_transaction() {
  assert(tx ! = nullptr);
  db->EndTransaction(*tx, [&](auto) {}) 
+ delete(this); // ADD THIS LINE
}
```
---

やっと理解できてきた気がします．的外れなコメントを繰り返していてすみません．
ただ，
> Ideally I want to delete LineairDBTransaction pointer (not LineairDB::Transaction) when I call end_transaction, but that seems impossible.

こちらのコメントの意味がわかりません．少なくとも `cleanup_*_tx()`  という関数を削除することはできると思います．
```c++
void LineairDBTransaction::end_transaction() {
  assert(tx != nullptr);
  db->EndTransaction(*tx, [&](auto) {}); 
+ delete this(); // ADD THIS LINE
}
```
このように `delete(this)` とすれば`cleanup_*_tx()` は不要なのではないでしょうか？

// 「Aの関数を呼ぶ前にかならずBを呼んでください」というコメントがあると，それは一つの関数にまとめるべきだと感じます．ので，お時間を頂いています．

Reply:
>I think we can do it and could delete the function cleanup_*_tx(). We can invoke delete(this). Is it not enough?
This solution worked, thank you for your suggestion and sorry for taking your time.